### ISSUE TITLE
Issue recasting array value to integer over 5 digits

### ISSUE THREAD

**Issue by RWUsask** [chapel-lang/chapel#7352]
### Summary of Problem
Passing array to subroutine, which are string values (but numbers)

In subroutine, if I attempt to recast the Array[x] values directly to int(64) variables, and the numeric value is over 9999, the operation fails with:
unexpected character "l" when converting from string to int(64)  **(lowercase L)

If I put the string value in an intermediate string variable, the operation works as expected.

### Steps to Reproduce
Array containing string values "1001","3004","10004" (or pick your own). If any value of Array is over 9999, error is observed.

Fails with error:
testSub(A) {
   var a1 int = A[1] int;
   var a2 int = A[2] int;
   var a3 int = A[3] int;   (ERROR HERE)
}

Works:
testSub(A) {
   var as1 string = A[1];
   var as2 string = A[2];
   var as3 string = A[3];
   var a1 int = as1 int;
   var a2 int = as2 int;
   var a3 int = as3 int;
}

**Compile command:**
chpl -o test test.chpl   (same with --fast)

**Execution command:**
./test

### Configuration Information
- Output of `chpl --version`: chpl Version 1.15.0
- Output of `$CHPL_HOME/util/printchplenv --anonymize`: bash: CHPL_TARGET_PLATFORM:: command not found
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: gcc (GCC) 7.2.0
- (For Cray systems only) Output of `module list`:
Platform: Linux64 on Arch Linux

**Comment by mppf** [comment]
Hi -

Sorry you are running in to an issue. The behavior does sound strange. I believe you're running in to a problem, but I'm unable to reproduce it. For one thing, it might be platform-specific or it might come about only when Chapel is configured in a particular way.

First, your code sample above doesn't quite use Chapel syntax. When making a bug report, it's important to include a complete example that fails, if possible. Maybe the colons got messed up when pasting it. So I tried making my own example, which works on master and on 1.15 for me:

``` chapel
proc testSub(A) {
  var a1 = A[1]:int;
  var a2 = A[2]:int;
  var a3 = A[3]:int;
  writeln(a1, " ", a2, " ", a3);
}


var arr = ["1001", "3004", "10004"];
testSub(arr);
```

Now, what worries me the most actually about this bug report is that the output of `$CHPL_HOME/util/printchplenv --anonymize` was `bash: CHPL_TARGET_PLATFORM:: command not found` - that would indicate a serious error and I wouldn't expect the compiler to build correctly if that's the case.

Can you check in to:
1. if `./util/printchpenv` works at all? Does it always give an error? Does it print anything else?
2. Does the program I pasted above work for you in the problematic configuration?
3. Does it work if you try again to follow the [Quickstart instructions](http://chapel.cray.com/docs/latest/usingchapel/QUICKSTART.html)?
4. Can you double-check that your example does not include an ell (`l`) character in the place of a one (`1`)? It can be hard to tell these apart (depending especially on the font you are using) and confusion about that would probably lead to the error you're reporting.

Thanks.

**Comment by RWUsask** [comment]
It appears that the printchplenv didn’t work as some issue with either the session I was running in or the cut & paste of the command from github. A manual type of the command in a new session yields:
$CHPL_HOME/util/printchplenv –anonymize
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: make
CHPL_ATOMICS: intrinsics
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none


I did try to make a shorter snippet (your example does work), so here is an example from my code:
(ZMQ client will send “bfs|44714|1009”)

// MAIN INPUT LOOP

var context: Context;
var socket = context.socket(ZMQ.REP);
socket.bind("tcp://*:5555");
writeln("Waiting for connections...");
var zmqcommand: string;
while zmqcommand != "quit" {
  var retmsg: string = "Exiting Mapper.";
  var msg = socket.recv(string);
  writeln("Recived ZMQ: ",msg);
  var Args = msg.split("|",2);
  zmqcommand = Args[1]: string;
  Args.remove(1);
  writeln("Command is: ",zmqcommand," Remaining args: ",Args);
  select zmqcommand {
     when "bfs" {
       retmsg = bfs(Args);
     }

  }
  socket.send(retmsg);
}

proc bfs (A) {
   writeln("Received: ",A[1]," and ",A[2]);
   var to = A[1]: int;     ## CRASH ON THIS LINE
   var from = A[2]: int;
   … do stuff and return BFS …
}


If my ZMQ client sends “bfs|9999|1009”, (or any 4-digit or less) value, this works beautifully, BFS path is returned to my client.  If I send anything 10,000 or above, I receive:
pdtest.chpl:120: error: Unexpected character when converting from string to int(64): 'l'

There are no warnings/errors during compile, this only happens during runtime.

If you need more code, I should be able to provide the whole source file.

-Rob

**Comment by mppf** [comment]
hi @RWUsask, since you're using ZMQ it's possible that the bug is in that module. Could you send a complete example, including code (whether in Chapel or no) to send to the ZeroMQ socket? I think your shorter snippet above is close. If you have trouble reproducing with that, maybe try running it all in valgrind? Thanks.

**Comment by RWUsask** [comment]
The ZMQ client (in its entirety) is here, it's mostly from an example on Chapel documentation page.

---

use ZMQ;

config var command: string = "bfs|44714|1009";

var context: Context;
var socket = context.socket(ZMQ.REQ);
socket.connect("tcp://localhost:5555");

socket.send(command);
var answer = socket.recv(string);
writeln("Received message from server: ", answer);

**Comment by cassella** [comment]
I see something wrong here too.  It seems to be related to the Args.remove().


```chapel
var msg = "11111111|44714|1009|1111|11111|111111";
var Args = msg.split("|");

config const rm = false;

if (rm) {
  Args.remove(1);
}
bfs(Args);


proc bfs(A) {
  writeln("Received: ", A);

  for a in A {
    writeln("And got ", a:int);
  }
}
```


```
fortytwo@magrathea:~/src/chpl$ ./notsobiguhoh -nl 1 
Received: 11111111 44714 1009 1111 11111 111111
And got 11111111
And got 44714
And got 1009
And got 1111
And got 11111
And got 111111
```

```
fortytwo@magrathea:~/src/chpl$ ./notsobiguhoh -nl 1 --rm
Received: 44714 1009 1111 11111 111111
And got 44714111
And got 10094
And got 1111
And got 111111
And got 111111
```

Note that three of the numbers in the second form do not match the
values printed as strings in the Received array.

**Comment by bradcray** [comment]
@casella: Thanks for boiling this down to a simple, non-ZMQ example.  I've verified that this bug was also present in 1.15
@RWUsask: Are you working from the master branch or from an official Chapel release?
@daviditen / @psahabu : I can't quite tell if this is a string bug or an array.remove() bug.  Would one of you have the time to triage it and own it?

**Comment by RWUsask** [comment]
@bradcray - I'm using official Chapel release (1.15.0)

**Comment by cassella** [comment]
@RWUsask I think you can work around this by writing the code in a way that doesn't use ``.remove()``.  For example, if I change my code to use ``var Args2 = Args[2..]`` instead, then calling ``bfs(Args2))`` works fine.  (On master at least; I haven't tried 1.15.0.)

**Comment by mppf** [comment]
I was curious so I investigated this bug. Here is what I found out.

The bug comes from a combination of several factors:
  * For historical reasons, string to int casting uses the null terminator at the end of the string (i.e. it's written with C strings). That's probably worth fixing on its own (i.e. make the c_string routines in runtime/src/chplcast.c take in a length argument).
  * Array.remove() works by assigning each element to the previous element.
  * The string assignment function (really in reinitString) wasn't setting the null terminator at the end of a string in certain cases.

Here is an even smaller program that shows the issue:
``` chapel
extern proc printf(fmt:c_string, arg:c_string);

var s1 = "0" * 10;
var s2 = "1" * 3;

var test = s1;
test = s2;

printf("%s\n", test.c_str());
```

In a minute I will make a PR that adjusts reinitString.
